WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift

```diff
@@ -0,0 +1,204 @@
+
+    init(site: Blog?,
+         domainPurchasedCallback: RegisterDomainCoordinator.DomainPurchasedCallback? = nil,
+         crashLogger: CrashLogging = .main) {
+        self.site = site
+        self.domainPurchasedCallback = domainPurchasedCallback
+        self.crashLogger = crashLogger
+    }
+
+    // MARK: Public Functions
+
+    func createCart(onSuccess: @escaping () -> (),
+                    onFailure: @escaping () -> ()) {
+        guard let domain else { return }
+        let siteID = site?.dotComID?.intValue
+        let proxy = RegisterDomainDetailsServiceProxy()
+        proxy.createPersistentDomainShoppingCart(siteID: siteID,
+                                                 domainSuggestion: domain.remoteSuggestion(),
+                                                 privacyProtectionEnabled: domain.supportsPrivacy ?? false,
+                                                 success: { _ in
+            onSuccess()
+        },
+                                                 failure: { _ in
+            onFailure()
+        })
+    }
+
+    func presentWebViewForNoSite(on viewController: UIViewController) {
+        guard let domain,
+              let url = URL(string: Constants.noSiteCheckoutWebAddress) else {
+            crashLogger.logMessage("Failed to present domain checkout webview for no-site.",
+                                   level: .error)
+            return
+        }
+
+        presentCheckoutWebview(on: viewController,
+                               domainSuggestion: domain,
+                               url: url,
+                               title: TextContent.checkoutTitle,
+                               shouldPush: true)
+    }
+
+    func presentWebViewForCurrentSite(on viewController: UIViewController) {
+        guard let domain,
+              let site,
+              let homeURL = site.homeURL,
+              let siteUrl = URL(string: homeURL as String), let host = siteUrl.host,
+              let url = URL(string: Constants.checkoutWebAddress + host) else {
+            crashLogger.logMessage("Failed to present domain checkout webview for current site.",
+                                   level: .error)
+            return
+        }
+
+        presentCheckoutWebview(on: viewController,
+                               domainSuggestion: domain,
+                               url: url,
+                               title: nil,
+                               shouldPush: false)
+    }
+
+    func handleNoSiteChoice(on viewController: UIViewController) {
+        createCart(
+            onSuccess: { [weak self] in
+                self?.presentWebViewForNoSite(on: viewController)
+            }) {
+                viewController.displayActionableNotice(title: TextContent.errorTitle, actionTitle: TextContent.errorDismiss)
+            }
+    }
+
+    func handleExistingSiteChoice(on viewController: UIViewController) {
+        print("handleExistingSiteChoice")
+    }
```

Comment:
Regarding my earlier [comment](https://github.com/wordpress-mobile/WordPress-iOS/pull/21979/files#r1385126656), I just saw these `handleNoSiteChoice` and `handleExistingSiteChoice`. So basically the coordinator has 5 public methods:

- createCart
- presentWebViewForNoSite
- presentWebViewForCurrentSite
- handleNoSiteChoice
- handleExistingSiteChoice

In my opinion, it is not clear when to use which method, and I think there is room for improvements here, for example:
- Cart creation is an implementation detail and should be called internally. Similar to what you did in `handleNoSiteChoice`. So I think `createCart` method should be `private`.
- `presentWebViewForNoSite` and `presentWebViewForCurrentSite` can also be private and called within `handleNoSiteChoice` and `handleExistingSiteChoice`.
- Maybe `handleNoSiteChoice` and `handleExistingSiteChoice` can be merged into a single public method and handle site vs no-site internally depending on nullability of `self.site`.

I'm not suggesting to make any changes right now, because that would break the existing "My Site > Domains" implementation. But I just wanted to hear you thoughts.

Reply:
Thanks for taking the time to make these thoughtful suggestions!

I do agree with most of this. And I have tried to address these suggestions in #21998, which builds on this PR, specifically in af5cf725a05bd07ccab9a18fda0bc60c45524ced. I did this to avoid causing conflicts.

I didn't merge `handleNoSiteChoice` and `handleExistingSiteChoice` because I think it would complicate things. But I applied your other suggestions.

Let me know what you think! Either here or in #21998 🙇

Reply:
> I didn't merge handleNoSiteChoice and handleExistingSiteChoice because I think it would complicate things. 

To clarify, there is what I meant by merging those 2 methods:

1. Mark `handleNoSiteChoice` and `handleExistingSiteChoice` as private.
2. Add a new public method that decides whether to call `handleExistingSiteChoice` or `handleNoSiteChoice` based on `self.site`

```swift

func handleDomainPurchasing() {
      if let site {
          handleExistingSiteChoice(site)
      } else {
           handleNoSiteChoice()
      }
}

private func handleNoSiteChoice() {
   ...
}

private func handleExistingSiteChoice(_ site: Blog) {
   ...
}
```

But I don't have a strong opinion on this, so feel free to decide whether you want make the change.